src/app/EventManagement.cpp

```diff
@@ -0,0 +1,903 @@
+            continue;
+        }
+
+        didWriteEvent = true;
+    }
+
+    {
+        // Check the number of bytes written.  If the event is too large
+        // to be evicted from subsequent buffers, drop it now.
+        CircularEventBuffer * buffer = mpEventBuffer;
+        do
+        {
+            VerifyOrExit(buffer->GetQueueSize() >= writer.GetLengthWritten(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+            if (buffer->IsFinalDestinationForPriority(opts.mpEventSchema->mPriority))
+            {
+                break;
+            }
+            else
+            {
+                buffer = buffer->mpNext;
+            }
+        } while (true);
+    }
+
+    mBytesWritten += writer.GetLengthWritten();
+
+exit:
+    ChipLogFunctError(err);
+    if (err != CHIP_NO_ERROR)
+    {
+        *mpEventBuffer = checkpoint;
+    }
+    else if (opts.mpEventSchema->mPriority >= CHIP_CONFIG_EVENT_GLOBAL_PRIORITY)
+    {
+        CircularEventBuffer * currentBuffer = GetPriorityBuffer(opts.mpEventSchema->mPriority);
+        aEventNumber                        = currentBuffer->VendEventNumber();
+        currentBuffer->AddEvent(opts.mTimestamp);
+
+#if CHIP_CONFIG_EVENT_LOGGING_VERBOSE_DEBUG_LOGS
+        ChipLogDetail(EventLogging,
+                      "LogEvent event number: %u schema priority: %u cluster id: 0x%x event id: 0x%x sys timestamp: 0x%" PRIx64,
+                      aEventNumber, opts.mpEventSchema->mPriority, opts.mpEventSchema->mClusterId, opts.mpEventSchema->mEventId,
+                      opts.mTimestamp.mValue);
+#endif // CHIP_CONFIG_EVENT_LOGGING_VERBOSE_DEBUG_LOGS
+
+        ScheduleFlushIfNeeded(apOptions == nullptr ? false : opts.mUrgent);
+    }
+
+    return err;
+}
+
+CHIP_ERROR EventManagement::CopyEvent(const TLVReader & aReader, TLVWriter & aWriter, EventLoadOutContext * apContext)
+{
+    TLVReader reader;
+    TLVType containerType;
+    CopyAndAdjustDeltaTimeContext context(&aWriter, apContext);
+    const bool recurse = false;
+    CHIP_ERROR err     = CHIP_NO_ERROR;
+
+    reader.Init(aReader);
+    err = reader.EnterContainer(containerType);
+    SuccessOrExit(err);
+
+    err = aWriter.StartContainer(AnonymousTag, kTLVType_Structure, containerType);
+    SuccessOrExit(err);
+
+    err = chip::TLV::Utilities::Iterate(reader, CopyAndAdjustDeltaTime, &context, recurse);
+    VerifyOrExit(err == CHIP_NO_ERROR || err == CHIP_END_OF_TLV, );
+
+    err = aWriter.EndContainer(containerType);
+    SuccessOrExit(err);
+
+    err = aWriter.Finalize();
+    SuccessOrExit(err);
+
+exit:
+    ChipLogFunctError(err);
+    return err;
+}
+
+CHIP_ERROR EventManagement::EventIterator(const TLVReader & aReader, size_t aDepth, void * apContext)
+{
+    CHIP_ERROR err     = CHIP_NO_ERROR;
+    const bool recurse = false;
+    TLVReader innerReader;
+    TLVType tlvType;
+    EventEnvelopeContext event;
+    EventLoadOutContext * loadOutContext = static_cast<EventLoadOutContext *>(apContext);
```

Comment:
Can we make the void* typesafe?

Reply:
CHIP_ERROR EventManagement::CopyEventsSince(const TLVReader & aReader, size_t aDepth, void * apContext) is used by  err = TLV::Utilities::Iterate(reader, CopyEventsSince, &context, recurse) in FetchEventSince function, Iterate function is https://github.com/project-chip/connectedhomeip/blob/master/src/lib/core/CHIPTLVUtilities.cpp#L62
I would create the follow-up PR to have the delegate pass to achieve typesafe, thanks

Reply:
https://github.com/project-chip/connectedhomeip/issues/6430